answer/variable.rs

```diff
@@ -6,16 +6,21 @@
 
 use std::fmt;
 
-use structural_equality::StructuralEquality;
+use structural_equality::{ordered_hash_combine, StructuralEquality};
 
 #[derive(Copy, Clone, Hash, Eq, PartialEq, Ord, PartialOrd)]
 pub struct Variable {
     id: VariableId,
+    anonymous: bool,
```

Comment:
This was the easiest way to make a non-anonymous and an anonymous variable not be structurally equal -- otherwise we end up with `match $_ has $_` and `match $x has $y;` as structurally equal.